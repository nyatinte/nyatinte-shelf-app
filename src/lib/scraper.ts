import { JSDOM } from 'jsdom';

export const scraper = async (url: string) => {
  const res = await fetch(url);
  const html = await res.text();
  const dom = new JSDOM(html);
  const doc = dom.window.document;
  const title = doc.querySelector('title')?.textContent;
  const description =
    doc.querySelector('meta[name="description"]')?.getAttribute('content') ??
    doc
      .querySelector('meta[property="og:description"]')
      ?.getAttribute('content');
  let image = doc
    .querySelector('meta[property="og:image"]')
    ?.getAttribute('content');

  if (image?.startsWith('/')) {
    const urlObj = new URL(url);
    image = `${urlObj.origin}${image}`;
  }

  let favicon =
    doc.querySelector('link[rel="icon"]')?.getAttribute('href') ??
    doc.querySelector('link[rel="shortcut icon"]')?.getAttribute('href');

  if (favicon?.startsWith('/')) {
    const urlObj = new URL(url);
    favicon = `${urlObj.origin}${favicon}`;
  }

  return {
    title,
    description,
    image,
    favicon,
  };
};
