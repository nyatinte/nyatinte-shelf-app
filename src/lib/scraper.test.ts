import { describe, expect, it } from 'bun:test'
import { scraper } from './scraper'

const testUrl = 'https://nandemo-eat.vercel.app/'
describe(scraper.name, async () => {
  it('should return the title, description, image, and favicon of the URL', async () => {
    const result = await scraper(testUrl)

    expect(result).toMatchObject({
      title: expect.any(String),
      description: expect.any(String),
      image: expect.any(String),
      favicon: expect.any(String),
    })
  })
  it('snapshot', async () => {
    const result = await scraper(testUrl)

    expect(result).toMatchSnapshot()
  })
})
