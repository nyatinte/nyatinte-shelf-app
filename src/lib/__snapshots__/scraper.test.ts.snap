// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`scraper snapshot 1`] = `
{
  "description": "なんでもEATは、｢ご飯何がいい？｣｢なんでもいいよ｣というやり取りを解決するサービスです。",
  "favicon": "https://nandemo-eat.vercel.app/favicon.ico",
  "image": "https://nandemo-eat.vercel.app/%E3%81%AA%E3%82%93%E3%81%A7%E3%82%82EAT.png",
  "title": "なんでもEAT",
}
`;
